### ISSUE TITLE
Simulation based integration testing

### ISSUE THREAD

**Issue by noseglasses** [keyboardio/Kaleidoscope#657]
This is a (possibly incomplete) but hopefully growing collection of firmware features and plugins that might require testing.

- [x] keymaps 
    - [x] standard keys (without modifiers and special stuff)
    - [x] keys with modifiers (and combinations of those)
    - [ ] shift to layer
    - [ ] layer toggle
    
- [ ] plugins
    - [ ] Cycle
    - [ ] CycleTimeReport
    - [ ] EEPROM-Keymap
    - [ ] EEPROM-Keymap-Programmer
    - [ ] EEPROM-Settings
    - [ ] Escape-OneShot
    - ~[ ] FirmwareDump (not possible with simulation)~
    - [ ] FocusSerial
    - [ ] GhostInTheFirmware
    - [ ] Leader
    - [ ] Macros
    - [ ] MagicCombo
    - [ ] MouseKeys
    - [ ] NumPad
    - [ ] OneShot
    - [ ] Qukeys
    - [ ] Redial
    - [ ] ShapeShifter
    - [ ] SpaceCadet
    - [ ] Syster
    - [ ] TapDance
    - [ ] TopsyTurvy
    - [ ] TypingBreaks
    - [ ] Unicode
    - [ ] WinKeyToggle
    
- [ ] LED effects
    - [ ] Colormap
    - [ ] FingerPainter
    - [ ] Heatmap
    - [ ] IdleLEDs
    - [ ] LED-ActiveLayerColor
    - [ ] LED-ActiveModColor
    - [ ] LED-AlphaSquare
    - [ ] LEDEffect-BootAnimation
    - [ ] LEDEffect-BootGreeting
    - [ ] LEDEffect-Breathe
    - [ ] LEDEffect-Chase
    - [ ] LEDEffect-Rainbow
    - [ ] LEDEffect-SolidColor
    - [ ] LED-Stalker
    - [ ] LED-Wavepool: *Attention: Possibly random behavior*
    
## About this page 

This page is meant to be a place to collect which parts of the firmware require testing and how this is best achieved.

Are there any plugins or features missing in the above list? Please let me know so I can add them. Also I can add more detail to the list if necessary.

Once Kaleidoscope's keyboard simulator is ready for use, the information collected on this page might become a starting point for setting up a simulation based integration test bench.

In the following I will try to explain in detail what will be possible with the upcoming simulator and how testing might work based on it.

## Current state of testing

Currently, Kaleidoscope relies on Travis CI to do compile testing of a set of test-sketches (smoke-exmples).

When a PR is submitted, a bunch of firmware sketches situated in the `example` directory tree of the core firmware repository are build automatically. A PR is typically refused if any of such builds fail. Due to the large variety of sketches, this likely covers all important aspects of the firmware.

## Simulation based testing

While the approach described above is very well suited for catching compile and link time errors, a firmware that builds without issues does not necessarily run correctly.

It would of course be cool to run-time test the entire firmware with all its bells and whistles on all supported keyboard hardwares, whenever a new PR is submitted. But due to hardware being involved, this is by no means possible. Fortunately, simulation based integration testing can take us pretty close to this goal.

The [keyboard simulator](https://github.com/CapeLeidokos/Kaleidoscope-Simulator) I am writing about runs the actual firmware. There are only few aspects of the real keyboard that cannot be simulated. Namely, the generation of real USB-reports and the exact timing of the hardware. Also, the simulated code is not binary compatible (nor is RAM and PROGMEM layout).

But apart from these minor restrictions, the virtual keyboard simulates the physical one rather realisticly.

## Testing

A Kaleidoscope keyboard can be modelled as a deterministic process as it relies in on deterministic algorithms in most parts.

Thus, the question of testing boils down to verifying that a defined sequence of actions (inputs) triggers an equally well defined sequence of reactions (outputs).

After an aspect that is to be tested has been isolated and it has been defined how test success is defined, the timing of the real process is turned into a C++ program that matches a test run. 

### Timing

For the real keyboard the duration of each cycle depends on the associated computational workload. That scales with the number of active plugins and their event handlers. Cycle duration will typically vary from cycle to cycle and will depend on the amount of user input. This makes cycle duration a measure that is hard to predict or estimate.

That's why in most cases (exceptions, see below) we will rely on fixed cycle duration as a rough model for reality. An estimate for cycle duration can be determined by running the simulated firmware on the real hardware under realistic conditions with respect to user interaction, and averaging cycle time measurements.

Yet another simplification of the simulation model is that time stands still within cycles. In the simulated process time is only allowed to jump between cycles. This stepwise increase decouples simulation wall time and simulated time. The speed of the machine that runs the simulator thus becomes a non-critical issue - which it wouldn't be anyway as the host, typically a x86, is orders of magnitude faster than the keyboard's MCU. For real-time simulations this even implies that the simulation process must be slowed down artificially to allow simulated time to be in sync with wall time.
 
Time standing still in each cycle is, however, acceptable as most firmware features query `millis()` only once a cycle, if at all. Great parts of the firmware even rely on a dedicated function to query time that caches the value of `millis()` at the beginning of each cycle.

Due to execution speed not being an issue, the simulator can safely do any additional computational work that it requires for testing, without affecting the behavior of the simulation by any means.

There is, however, one drawback of fixed cycle times. They do prevent detection of specific timing issues. To provoke such issues or to verify that there are non of them, the simulator can be made to vary cycle times in a deterministic pseudo-random fashion.

### Types of action

There are two different types of action that cause changes of the (internal) state of the keyboard or trigger reactions. Namely, key action (pressing and releasing keys) and serial communication.

#### Key action

Key action, i.e. changes of the keys' states (pressed/release) are considered only between simulation cycles. This simplification in comparison to the real process is  justified as the time it takes to scan the key matrix at the beginning of each cycle is comparably short compared to the time scale on which user interaction takes place. The simulator passes the key states it stores to the firmware during key matrix scans, thereby it mimics a real key scanner.

#### Serial communication

Serial communication is more complex than key action as the stream of data that is send and received is less standardized. Serial communication is simulated by using two character queues for I/O. Both are written to and read from in between between cycles to mimic the exchange of information with the host. As serial communication is comparably slow, it is acceptable that the firmware sees no reactions to any of its requests within the same cycle.

### Verifying reactions

With integration testing, we treat the keyboard as a black box and concentrate on reactions rather than on what happens to the internal states of the keyboard.

By defining how we expect actions to cause reactions rather than caring for implementation details, in an optimal case we allow for future evolution of the firmware without having to adapt any tests.

The question about what is an internal state and what is not might not in all cases be answered easily. But there are aspects where this is beyond dispute. USB reports, e.g., are a typical reactions and thus well suited for integration testing.

Although stateful, verifying LED states is also a good candidate for integration testing as most LED effects precisely define what they intent to display.

The state of RAM, in contrast, is clearly a bad criterium for testing. Test cases that check the values of (internal) state variables tend to be very sensitive to changes of the design of a program and are, thus, better avoided.

#### Verifying keyboard and mouse reports

For the verification of keyboard/mouse reports, it is important to precisely define at which point in time a report is expected (or not expected) and what keycodes it must contain (or must not).  As multiple reports can occur within the same cycle, their precise order must also be defined and verified.

### Verifying EEPROM and LEDs

Verifying the state of EEPROM may be suitable for automatic testing as EEPROM layout is somewhat standardized by most related plugins.

The state of both EEPROM and LEDs can be tested conveniently by generating a raw dump of the simulators arrays. Such data, generated during reference runs, can later be used for automatic comparisons with the actual state of the respective variables.

**Comment by gedankenexperimenter** [comment]
I think we can safely close this issue, given that we now have a working test simulator and a fairly broad test suite.
